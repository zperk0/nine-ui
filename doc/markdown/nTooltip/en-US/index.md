# Tooltip
Commonly used to display the prompt information when the mouse hover.
## When To Use
Used for auxiliary text prompts, which can replace the default title attribute of HTML elements.

## Examples

### Basic Use
:::demo
```html
<template>
  <n-tooltip content="Tooltip some content" placement="top">
    <n-button>Default</n-button>
  </n-tooltip>
</template>
```
:::

## API
| Property | Description | Type | Default |
| :--- | :--- | :--- | :--- |
| content | Tooltip content | String |  |
| placement | position | String | `top` |
| trigger | Trigger event type | String | `hover` |