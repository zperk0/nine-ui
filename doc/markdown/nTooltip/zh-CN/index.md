# Tooltip 文字提示
常用于展示鼠标 hover 时的提示信息。
## 何时使用
用于辅助的文字提示，可代替 HTML 元素默认的 title 属性。

## 代码演示

### 基本使用
:::demo
```html
<template>
  <div class="tooltip-example">
    <div class="top">
      <n-tooltip class="item" content="Tooltip 文字提示" placement="top">
        <n-button>上边</n-button>
      </n-tooltip>
    </div>
    <div class="left">
      <n-tooltip class="item" content="Tooltip 文字提示" placement="left" trigger="click">
        <n-button>左边</n-button>
      </n-tooltip>
    </div>
    <div class="right">
      <n-tooltip class="item" content="Tooltip 文字提示" placement="right" trigger="click">
        <n-button>右边</n-button>
      </n-tooltip>
    </div>
    <div class="bottom">
      <n-tooltip class="item" content="Tooltip 文字提示" placement="bottom">
        <n-button>下边</n-button>
      </n-tooltip>
    </div>
  </div>
</template>
```
:::

## API

| 参数 | 说明 | 类型 | 默认值 |
| :--- | :--- | :--- | :--- |
| content | 提示内容 | String |  |
| placement | 位置 | String | `top` |
| trigger | 触发事件类型 | String | `hover` |