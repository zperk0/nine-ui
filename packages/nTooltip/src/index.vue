<template>
  <div class="n-tooltip">
    <div
      ref="popover"
      class="tooltip-container"
      role="tooltip"
      :class="[ show ? 'visible' : 'not-visible']"
    >
      <div class="content">
        {{ content }}
      </div>
      <div
        class="arrow"
        :class="placement"
      />
    </div>
    <span
      ref="trigger"
    >
      <slot />
    </span>
  </div>
</template>

<script>
import EventListener from '../../../src/utils/eventlistener';
export default {
  name: 'NTooltip',
  props: {
    trigger: {
      type: String,
      default: 'hover'
    },
    content: {
      type: String,
      default: ''
    },
    placement: {
      type: String,
      default: 'top'
    }
  },
  data () {
    return {
      position: {
        top: 0,
        left: 0
      },
      show: false
    };
  },
  watch: {
    show: function (val) {
      if (val) {
        const popover = this.$refs.popover;
        const trigger = this.$refs.trigger.children[0];
        switch (this.placement) {
          case 'top' :
            this.position.left = trigger.offsetLeft - popover.offsetWidth / 2 + trigger.offsetWidth / 2;
            this.position.top = 0 - trigger.offsetHeight;
            break;
          case 'left':
            this.position.left = trigger.offsetLeft - popover.offsetWidth;
            this.position.top = trigger.offsetTop + trigger.offsetHeight / 2 - popover.offsetHeight / 2;
            break;
          case 'right':
            this.position.left = trigger.offsetLeft + trigger.offsetWidth;
            this.position.top = trigger.offsetTop + trigger.offsetHeight / 2 - popover.offsetHeight / 2;
            break;
          case 'bottom':
            this.position.left = trigger.offsetLeft - popover.offsetWidth / 2 + trigger.offsetWidth / 2;
            this.position.top = trigger.offsetTop + trigger.offsetHeight;
            break;
        }
        this.$refs.popover.style.top = this.position.top + 'px';
        this.$refs.popover.style.left = this.position.left + 'px';
      }
    }
  },
  mounted () {
    const trigger = this.$refs.trigger.children[0];
    if (trigger) {
      if (this.trigger === 'hover') {
        this._mouseenterEvent = EventListener.listen(trigger, 'mouseenter', () => {
          this.show = true;
        });
        this._mouseleaveEvent = EventListener.listen(trigger, 'mouseleave', () => {
          this.show = false;
        });
      } else if (this.trigger === 'focus') {
        this._focusEvent = EventListener.listen(trigger, 'focus', () => {
          this.show = true;
        });
        this._blurEvent = EventListener.listen(trigger, 'blur', () => {
          this.show = false;
        });
      } else {
        this._clickEvent = EventListener.listen(trigger, 'click', this.toggle);
      }
    }
  },
  methods: {
    toggle () {
      this.show = !this.show;
    }
  }
};
</script>

<style lang="scss">
  .n-tooltip{
    position: relative;
    .tooltip-container{
      position: absolute;
      font-family: PingFangSC-Regular;
      font-size: 12px;
      color: #FFFFFF;
      line-height: 24px;
      max-width:250px;
      .content{
        display: inline-block;
        background: #272C36;
        border-radius: 4px;
        min-width:104px;
        min-height:24px;
        width:auto;
        color:#ffffff;
        padding:0 10px;
      }
      .arrow{
        position: relative;
        width: 0;
        height: 0;
        border-top: 6px solid #272C36;
        border-right: 6px solid transparent;
        border-left: 6px solid transparent;
      }
      .top{
        margin: 0 auto;
        text-align:center;
      }
    }
    .visible{
      visibility: visible;
    }
    .not-visible{
       visibility: hidden;
    }
  }
</style>
