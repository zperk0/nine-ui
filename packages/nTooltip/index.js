import nTooltip from './src/index.vue';

nTooltip.install = function (Vue) {
  Vue.component(nTooltip.name, nTooltip);
};

export default nTooltip;
