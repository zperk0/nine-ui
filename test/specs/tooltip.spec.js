import nTooltip from '../../packages/n-tooltip';
import { mount } from '@vue/test-utils';

describe('nTooltip', () => {
  it('should return NTooltip', () => {
    const wrapper = mount({
      render () {
        return <nTooltip />;
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
