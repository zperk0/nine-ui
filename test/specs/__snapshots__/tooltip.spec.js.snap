// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nTooltip should return NTooltip 1`] = `
<div class="n-tooltip">
  <div class="tooltip-container top not-visible" role="tooltip">
    <div class="content"></div>
    <div class="arrow top"></div>
    <!---->
  </div><span><!----></span>
</div>
`;
